db/database.go

```diff
@@ -85,18 +112,76 @@ func (db *Database) ClearAllOffers() error {
 	defer iter.Release()
 
 	for iter.Valid() {
-		// key is the offer ID
+		offerID := iter.Key()
+		err := db.offerTable.Del(offerID)
+		if err != nil {
+			return err
+		}
+		iter.Next()
+	}
+
+	return nil
+}
+
+// PutSwap puts the given swap in the database.
+// If a swap with the same ID is already in the database, it overwrites it.
+func (db *Database) PutSwap(s *swap.Info) error {
+	val, err := json.Marshal(s)
+	if err != nil {
+		return err
+	}
+
+	key := s.ID
+	return db.swapTable.Put(key[:], val)
+}
+
+// HasSwap returns whether the db contains a swap with the given ID.
+func (db *Database) HasSwap(id types.Hash) (bool, error) {
+	return db.swapTable.Has(id[:])
+}
+
+// GetSwap returns a swap with the given ID, if it exists.
+func (db *Database) GetSwap(id types.Hash) (*swap.Info, error) {
+	value, err := db.swapTable.Get(id[:])
+	if err != nil {
+		return nil, err
+	}
+
+	var s swap.Info
+	err = json.Unmarshal(value, &s)
+	if err != nil {
+		return nil, err
+	}
+
+	return &s, nil
+}
+
+// GetAllSwaps returns all swaps in the database.
+func (db *Database) GetAllSwaps() ([]*swap.Info, error) {
+	iter := db.swapTable.NewIterator()
+	defer iter.Release()
+
+	var swaps []*swap.Info
+	for iter.Valid() {
 		key := iter.Key()
-		if len(key) != 32 {
-			panic("key (offer ID) length is not 32")
+
+		// if the key becomes longer than 32, we're not iterating over swaps
+		if len(key) > idLength {
+			break
 		}
 
-		err := db.offerTable.Del(key)
+		// value is the encoded swap
+		value := iter.Value()
+
+		var s swap.Info
+		err := json.Unmarshal(value, &s)
 		if err != nil {
-			return err
+			return nil, err
 		}
+
+		swaps = append(swaps, &s)
```

Comment:
Are there any other sanity checks that we should do besides the unmarshal? Besides data corruption, there is also the potential issue of the data we storing changing with new releases.

Reply:
hm, like if the `swap.Info` type changes? I guess in that case, the unmarshal should fail (probably). do you think adding a version byte would be good?

Reply:
I was thinking we might need a semantic version as the first field.  If the major version changed, we'd have to use a new struct, because it was a backwards incompatible change.  We could only support one previous major version to keep things simple.  We would throw an error if we discovered a version newer than the latest version that the current `swapd` supports. Minor versions could be increased if you are just adding a field, but the older version would still be compatible. (That's off the top of my head, it may not be perfect.)

Reply:
Not saying that all needs to be done this pull request, but we will want it before we get too many users.